src/db.c

```diff
@@ -1918,94 +1917,103 @@ int xreadGetKeys(struct redisCommand *cmd, robj **argv, int argc, getKeysResult
+    dictEntry *prev = dictEntryPrevInSlot(entry);
+    if (next != NULL) {
+        dictEntryPrevInSlot(next) = entry;
+    }
+    if (prev != NULL) {
+        dictEntryNextInSlot(prev) = entry;
     } else {
-        raxFree(rt);
+        /* The replaced entry was the first in the list. */
+        sds key = entry->key;
+        unsigned int hashslot = keyHashSlot(key, sdslen(key));
+        server.cluster->slots_to_keys[hashslot] = entry;
     }
 }
 
-/* Empty the slots-keys map of Redis CLuster by creating a new empty one and
- * freeing the old one. */
-void slotToKeyFlush(int async) {
-    rax *old = server.cluster->slots_to_keys;
-
-    server.cluster->slots_to_keys = raxNew();
-    memset(server.cluster->slots_keys_count,0,
+/* Empty the slots-keys map of Redis Cluster. */
+void slotToKeyFlush(void) {
+    memset(server.cluster->slots_to_keys, 0,
+           sizeof(server.cluster->slots_to_keys));
+    memset(server.cluster->slots_keys_count, 0,
            sizeof(server.cluster->slots_keys_count));
-    freeSlotsToKeysMap(old, async);
 }
 
 /* Populate the specified array of objects with keys in the specified slot.
  * New objects are returned to represent keys, it's up to the caller to
  * decrement the reference count to release the keys names. */
 unsigned int getKeysInSlot(unsigned int hashslot, robj **keys, unsigned int count) {
-    raxIterator iter;
-    int j = 0;
-    unsigned char indexed[2];
-
-    indexed[0] = (hashslot >> 8) & 0xff;
-    indexed[1] = hashslot & 0xff;
-    raxStart(&iter,server.cluster->slots_to_keys);
-    raxSeek(&iter,">=",indexed,2);
-    while(count-- && raxNext(&iter)) {
-        if (iter.key[0] != indexed[0] || iter.key[1] != indexed[1]) break;
-        keys[j++] = createStringObject((char*)iter.key+2,iter.key_len-2);
-    }
-    raxStop(&iter);
+    unsigned int j = 0;
+    dictEntry *de = server.cluster->slots_to_keys[hashslot];
+    while (j < count && de != NULL) {
+        sds sdskey = dictGetKey(de);
+        keys[j++] = createStringObject(sdskey, sdslen(sdskey));
+        de = dictEntryNextInSlot(de);
+    }
     return j;
 }
 
 /* Remove all the keys in the specified hash slot.
  * The number of removed items is returned. */
 unsigned int delKeysInSlot(unsigned int hashslot) {
-    raxIterator iter;
-    int j = 0;
-    unsigned char indexed[2];
-
-    indexed[0] = (hashslot >> 8) & 0xff;
-    indexed[1] = hashslot & 0xff;
-    raxStart(&iter,server.cluster->slots_to_keys);
-    while(server.cluster->slots_keys_count[hashslot]) {
-        raxSeek(&iter,">=",indexed,2);
-        raxNext(&iter);
-
-        robj *key = createStringObject((char*)iter.key+2,iter.key_len-2);
-        dbDelete(&server.db[0],key);
+    unsigned int j = 0;
+    dictEntry *de = server.cluster->slots_to_keys[hashslot];
+    while (de != NULL) {
+        sds sdskey = dictGetKey(de);
+        de = dictEntryNextInSlot(de);
+        robj *key = createStringObject(sdskey, sdslen(sdskey));
+        dbDelete(&server.db[0], key);
```

Comment:
maybe we could have used stack based string object (wrapping the existing sds)?
i suppose it can cause issues for modules that will get a notification with that stack based object (unless they already use the new RM_HoldString)
WDTY?

Reply:
Maybe, but it seems to be some work to investigate it. I'll have a quick look. The old code also created an robj from an sds key, so this would be an orthogonal change, i.e. probably leave it for another PR.

Reply:
yeah i noticed that was already the case, and as i noted, the change carries some implications.
i suppose we better leave that for another PR anyway (because of the implications).